### ISSUE TITLE
Decompile relocatable file (powerpc)

### ISSUE THREAD

**Issue by Leseratte10** [avast-tl/retdec#67]
I have a small powerpc application, which has most of its actual logic in a relocatable file and links that during runtime, similar as a DLL on Windows. 

http://www.metroid2002.com/retromodding/wiki/REL_(File_Format) 
http://wiki.tockdom.com/wiki/REL_(File_Format) 

Is there any way retdec can decompile this REL file, too? It contains mostly code. 
Can I tell retdec to force decompile this as raw file, treating everything as assembler code? 

`./decompile.sh -a powerpc -e big -f raw ../../compil/StaticR.rel` tells me "raw" would be an unsupported format? Why? According to the readme, retdec supports raw machine code ...

**Comment by s3rvac** [comment]
Hi. You have a good point about `-f raw`. Currently, we do not treat raw machine code as a file format, but rather as a decompilation mode (`--mode raw` or simply `-m raw`). Thus, to decompile the input file as raw machine code, you will need to run
```
./decompile.sh -a powerpc -e big -m raw --raw-entry-point ADDR1 --raw-section-vma ADDR2 ../../compil/StaticR.rel
```
where `ADDR1` and `ADDR2` are hexadecimal addresses (e.g. `0x10000300`). We do not yet have a description of raw decompilation on our wiki, but you can read our [raw decompilation guide](https://retdec.com/api/docs/guides.html#raw-decompilation) that we wrote for users of [retdec.com's API](https://retdec.com/api/). Feel free to contact us if you have any problems with the raw decompilation.

**Comment by Leseratte10** [comment]
Hm, this is a relocatable module, so it provides functions to a main binary. It does have a VMA address (although I don't know that one yet), but as far as I know, it has no entry point - or does it?

**Comment by s3rvac** [comment]
Good point. @PeterMatula, is it possible to decompile such a file via raw decompilation, or will this require additional support in the decompiler?

**Comment by Leseratte10** [comment]
I have now tried a complete decompilation of the system's RAM dump (~25 MB) and it crashed with the following error: 

```
Running phase: Initialization ( 0.01s )
Running phase: LLVM ( 0.01s )
Running phase: Providers initialization ( 0.01s )
Running phase: Input binary to LLVM IR decoding ( 0.63s )
Running phase: Assembly instruction optimization ( 28.46s )
Running phase: LLVM ( 29.55s )
Running phase: (Un)Volatilize optimization ( 32.39s )
Running phase: LLVM ( 33.79s )
Running phase: (Un)Volatilize optimization ( 54.62s )
Running phase: Control flow optimization ( 56.26s )
Running phase: Control flow function detection optimization ( 7026.86s )
Running phase: Main function identification optimization ( 7041.03s )
Running phase: Assembly register optimization ( 7041.06s )
Running phase: Stack optimization ( 7041.06s )
Running phase: Control flow optimization ( 7079.73s )
Running phase: Conditional branch optimization ( 7102.75s )
Running phase: Syscalls optimization ( 7120.63s )
Running phase: Libgcc idioms optimization ( 7120.63s )
Running phase: Constants optimization ( 7120.63s )
Running phase: Function parameters and returns optimization ( 7603.03s )
Running phase: Register localization optimization ( 7621.99s )
Running phase: Data type conversions optimization ( 7637.42s )
Running phase: Simple types recovery optimization ( 7638.12s )
terminate called after throwing an instance of 'St9bad_alloc'
  what():  std::bad_alloc
./decompile.sh: Zeile 900: 23337 Abgebrochen             (Speicherabzug geschrieben) $BIN2LLVMIR "${BIN2LLVMIR_PARAMS[@]}" -o "$OUT_BACKEND_BC"
Error: Decompilation to LLVM IR failed
```
Is that a bug caused by the file? Is it because it is so large?

I'll try decompiling just the REL file next ...

**Comment by s3rvac** [comment]
The error means that the decompilation process failed to allocate the needed memory (you ran out of memory). Nevertheless, judging from the runtime (more than 7 000 seconds), the file is either too big to be decompiled in this way, or its instructions were incorrectly decoded in the `Input binary to LLVM IR decoding` phase. I think it will be better to try decompiling the original file than a RAM dump.

**Comment by PeterMatula** [comment]
Few notes:
1. ~25 MB file is currently just to much for `retdec`, I would not expect the decompilation to end with some useful output even if it was not PPC code in an unsupported file format.
2. In theory, `raw` mode can do something about it, if you set correct values to give it something to start with. But again, in practice, I don't think anything useful will be produced at the moment. There are big issues that we need to fix even in PE/ELF x86 relocatables, REL PPC will not be any better (much worse i presume).
3. If there is some known container for the data (REL in this case), it would be the best to implement its proper handling in `fileformat` library. Even if there is not much to parse, it is better than `raw` mode, which is an unreliable last resort. It should not be that hard, and it might be a good starting point for someone that would like to help with the development. Proper parsing can be tested just by analyzing files with `fileinfo` utility, at this point, decompilation does not even need to be considered.
4. To improve the decompilation output, the next thing should be experimentation with REL files and identification what is wrong with the outputs, why, and what can be done about it. Even though we try to make `retdec` general and not limited (or as little as possible) to object files, architectures, etc., testing and small tweaks are usually needed to get some good quality outputs for this kind of cases (entirely new class of inputs that we never tested and optimized for them before).
